### DISCUSSION TITLE
How to send csv report as an attachment

### DISCUSSION THREAD

**Discussion by pricy** [cloud-custodian/cloud-custodian#7352] [General]
I have a tag compliance policy but need to send the violation report as a CSV file attachment instead of json within the email body. How do I do that?

I can generate the csv report using this command but how can I default to this instead of sending out the resources.json output?

custodian report --output-dir=. --format csv --field tags=tags --region us-east-1 policy.yml > ec2_tag_compliance.csv

**Comment by kapilt** [comment]
so mailer doesn't really support attachments, so the best option is using invoke-lambda action or a sns delivery target on notify. note any data payload is effectively capped at service transport limits re size (256kb commonly) for which services are chunked over multiple invocations for a given policy execution.

**Comment by pricy** [comment]
Thank you.

**Comment by nitrocode** [comment]
Is it planned to have the mailer support attachments? I'd imagine, the policy could

1. output resources json to s3
2. in the notify block, pass in the fields and jmespath used for the `custodian report` command
3. mailer run the same `custodian report` code against the resources.json
4. upload the output directly to another s3 output location
5. email the link to the S3 location

```
policies:
  - name: root-user-login-detected
    resource: account
    description: |
      Notifies Security and Cloud Admins teams on any AWS root user console logins
    mode:
      type: cloudtrail
      events:
        - ConsoleLogin
      execution-options:
        output_dir: s3://{account_id}-custodian/output
      role: snip
    filters:
      - type: event
        key: "detail.userIdentity.type"
        value: Root
    actions:
      - type: notify
        to:
          - user@company.com
        report:
          output_dir: s3://{account_id}-custodian/reports
          default_names: false
          fields:
            LaunchTime: LaunchTime
            State: State.Name
            Name: tag:Name
            Environment: tag:env
        transport:
          type: sqs
          queue: https://sqs.us-west-2.amazonaws.com/{account_id}/cloud-custodian
          region: us-east-1
```

thoughts? @kapilt

**Comment by kapilt** [comment]
this feels like something that should be handled in mailer, not in the notify action, ie its a presentation concern, and the mailer is responsible for formatting messages for delivery. it could be as simple as a template function {{ attach_csv() }}.
